### PR TITLE
Fixing the UBSan warnings

### PR DESCRIPTION
**Pull request by myrrc**
Building this library with UBSan results in a bug:

```
impl.h:188:41: runtime error: 9.22337e+18 is outside the range of representable values of
type 'long'
```

The reason is that some double values may not fit in int64, although the initial values could.
See also https://stackoverflow.com/questions/61877751/outside-the-range-of-representable-values-when-casting-double-to-long-long-with

One possible solution here is to explicilty check for the values using compiler-provided __int128 and, if found overflow, explicitly add the non-overflowing parts.

### CODE REVIEW

File: v3/impl.h
Comment (apolukhin) [inline v3/impl.h]: __int128_t is an unportable extension. Will `(r - to_diff) > static_cast<long double>(max_int)` do the job? You may add a static assert, that long double has enough bits

Comment (myrrc) [reply v3/impl.h]: @apolukhin Looks like the long double would do on most systems, and now there's a static assertion for the others.

**Comment by myrrc** [comment]
Note that the second UB could lead to incorrect calculations, consider the report https://clickhouse-test-reports.s3.yandex.net/15419/b7e716fb2b993218c14568c9876e95a0a84335e9/functional_stateless_tests_flaky_check_(address)/test_run.txt.out.log

(the "wanted result" (reference) is before the fix -- the last column's sign is wrong).

File: v3/impl.h
Comment (cerevra) [inline v3/impl.h]: AFAIK `size_t` is used here as alias for `uint64_t`.
`size_t` looks like harmful type here: it should be replaced with `uint64_t` if it is possible. Different definitions of `size_t` and `unit64_t` could lead to some problems.

File: v3/impl.h
Comment (cerevra) [inline v3/impl.h]: Could you explain please: there is division by `max_int` in L231 but later `self` would be multiplied by `max_uint` in L239.
Why there is the difference?

MAYBE all `uint64_t` and `size_t` should be replaced with `int64_t`? Probably all type mismatches would be fixed without problems? Or not?
@apolukhin @myrrc

**Review by cerevra — CHANGES_REQUESTED** [review]

File: v3/impl.h
Comment (myrrc) [reply v3/impl.h]: I personally don't see any pros of using size_t over uint64_t here, let's ask @apolukhin

File: v3/impl.h
Comment (myrrc) [reply v3/impl.h]: > Could you explain please: there is division by `max_int` in L231 but later `self` would be multiplied by `max_uint` in L239.
> Why there is the difference?
> 
> MAYBE all `uint64_t` and `size_t` should be replaced with `int64_t`? Probably all type mismatches would be fixed without problems? Or not?
> @apolukhin @myrrc

There are some nasty cases as I've described before (rounding up the values of int64 and uint64 when converting them to double may lead to disability to convert them back to uint64/int64), so this question needs a more thorough review.

File: v3/impl.h
Comment (cerevra) [inline v3/impl.h]: This loop could be extremely slow.

Maybe `std::ilogb()`+`std::scalbn()` could help to fill `self` with some numerable iterations?

**Review by cerevra — CHANGES_REQUESTED** [review]

File: v3/impl.h
Comment (myrrc) [reply v3/impl.h]: This still is not an optimal solution, I prefer to think of it as a more-or-less working one.

Comment (myrrc) [reply v3/impl.h]: Turns out this while inappropriately slowed down everything.

**Comment by cerevra** [comment]
LGTM
Is there any other cases which should be covered before merging?

**Comment by myrrc** [comment]
> LGTM
> Is there any other cases which should be covered before merging?

Let's wait till the related ClickHouse PR is merged, maybe I'll find other bugs.

**Comment by myrrc** [comment]
@cerevra Looks like there are some issues about the ClickHouse PR review, so we'd better not wait.
I couldn't find any more bugs.

**Pull request merged**
